<launch>
  <arg name="input" default="/kinect_head/depth_registered/points" />
  <arg name="manager" default="tabletop_object_detector_nodelet_manager" />
  <arg name="machine" default="localhost" />

  <arg name="launch_manager" default="true" />
  <arg name="launch_tracking" default="true" />
  <arg name="launch_openni" default="true" />
  <arg name="launch_rviz" default="true" />
  <arg name="publish_tf" default="true" />
  
  <!-- parameter for box filter -->
  <arg name="box_filter_frame" default="base_link"/>

  <!-- parameter for hsi filter -->
  <arg name="h_max" default="8" />
  <arg name="h_min" default="-128" />
  <arg name="s_max" default="255" />
  <arg name="s_min" default="45" />
  <arg name="i_max" default="255" />
  <arg name="i_min" default="0" />

  <!-- parameter for euclidean clustering -->
  <arg name="clustering_tolerance" default="0.01"/>


  <machine name="localhost" address="localhost" />

  <!-- tabletop object detection -->
  <group ns="vase_detection">

  <node name="$(arg manager)" pkg="nodelet" type="nodelet" args="manager"
        machine="$(arg machine)" if="$(arg launch_manager)"
        output="screen" />

  <node name="input_relay" pkg="nodelet" type="nodelet"
        args="standalone jsk_topic_tools/Relay"
        machine="$(arg machine)">
    <remap from="~input" to="$(arg input)" />
  </node>

  <node name="boxfilter_pre" pkg="nodelet" type="nodelet"
    args="standalone jsk_pcl/AttentionClipper"
    output="screen">
    <remap from="~input/points" to="input_relay/output"/>
    <rosparam>
      initial_pos: [1, 0.0, 1]
      dimension_x: 2
      dimension_y: 1
      dimension_z: 1
      use_multiple_attention: false
    </rosparam>
    <param name="frame_id" value="$(arg box_filter_frame)" />
  </node>

  <node name="ExtractIndices" pkg="nodelet" type="nodelet"
    args="standalone jsk_pcl/ExtractIndices"
    output="screen">
    <remap from="~input" to="input_relay/output"/>
    <remap from="~indices" to="boxfilter_pre/output/point_indices"/>
    <rosparam>
      keep_organized: true
    </rosparam>
  </node>

  <node name="multi_plane_estimate" pkg="nodelet" type="nodelet"
        args="load jsk_pcl/OrganizedMultiPlaneSegmentation $(arg manager)"
        output="screen"
        if="$(arg launch_manager)"
        machine="$(arg machine)">
    <remap from="~input" to="ExtractIndices/output" />
    <rosparam>
      max_curvature: 0.01
      estimate_normal: true
    </rosparam>
  </node>

  <node name="multi_plane_extraction" pkg="nodelet" type="nodelet"
    args="load jsk_pcl/MultiPlaneExtraction $(arg manager)"
    output="screen"
    if="$(arg launch_manager)"
    machine="$(arg machine)">
    <remap from="~input" to="ExtractIndices/output"/>
    <remap from="~input_polygons" to="multi_plane_estimate/output_polygon"/>
    <remap from="~input_coefficients" to="multi_plane_estimate/output_coefficuents"/>
    <remap from="~indices" to="multi_plane_estimate/output"/>
    <rosparam>
      min_height: 0.01
      max_height: 0.6
    </rosparam>
  </node>

  <node name="hsi_filter" pkg="nodelet" type="nodelet"
        args="load jsk_pcl/HSIColorFilter $(arg manager)" output="screen">
    <remap from="~input" to="multi_plane_extraction/output" />
    <rosparam>
      use_indices: false
      keep_organized: false
    </rosparam>
    <param name="h_limit_max" value="$(arg h_max)" />
    <param name="h_limit_min" value="$(arg h_min)" />
    <param name="s_limit_max" value="$(arg s_max)" />
    <param name="s_limit_min" value="$(arg s_min)" />
    <param name="i_limit_max" value="$(arg i_max)" />
    <param name="i_limit_min" value="$(arg i_min)" />
  </node>

  <node pkg="nodelet" type="nodelet" name="euclidean_clustering"
        args="load jsk_pcl/EuclideanClustering $(arg manager)" output="screen">
    <remap from="~input" to="hsi_filter/output" />
    <rosparam>
      tolerance: 0.02
      min_size: 50
    </rosparam>
    <param name="tolerance" value="$(arg clustering_tolerance)"/>
  </node>

  <node pkg="nodelet" type="nodelet"
        name="cluster_decomposer"
        args="load jsk_pcl/ClusterPointIndicesDecomposerZAxis $(arg manager)"
        output="screen" clear_params="true">
    <remap from="~input" to="hsi_filter/output" />
    <remap from="~target" to="euclidean_clustering/output" />
    <remap from="~debug_output" to="debug_output" />
    <remap from="~boxes" to="boxes" />
    <rosparam>
    </rosparam>
  </node>

  <node pkg="jsk_interactive_marker" type="bounding_box_marker"
        name="bounding_box_marker">
    <remap from="~bounding_box_array" to="boxes" />
  </node>

  </group>

</launch>
